DM-36794: Initial version

rra opened the pull request.

rra committed:
Initial version

rra committed:
Install fonts-dejavu for blockdiag

rra committed:
Fix link to PyYAML

rra committed:
Add testing and a note about pathlib

rra committed:
Reword docstring coding style point

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+===============================
+
+In general, use Safir_ whenever it provides necessary functionality, and use whatever underlying libraries it supports.
+This includes HTTPX_ for HTTP clients and structlog_ for logging.
```
Also [arq](https://arq-docs.helpmanual.io).

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+
+- Write docstrings following the `Rubin project recommendations <https://developer.lsst.io/python/numpydoc.html>`__.
+  (As of this writing, this guide does not yet recommend omitting types from the docstrings, which is now better style when using current Sphinx.
+  This is likely to be fixed soon.)
```
```suggestion
  This is likely to be updated soon.)
```

(one of my pet peeves is classifying lack of a feature as a "bug" :) )

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+
+.. code-block:: python
+
+   @dataclass
```
I've been long-thinking that the app template should switch to Pydantic BaseSettings since it does all the work needed to convert string-based environment variables into appropriate Python types (and validate).

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+
+Do not mix the two approaches, since that can be quite confusing.
+When using YAML for configuration, get all of the configuration from the YAML file and not from environment variables.
+(A small number of special exceptions can be made if there are specific settings that need to be easily overridden for CI.)
```
Hm, I do break this rule in noteburst. I'll have an environment variable specify the path to a configuration file and then use that YAML configuration file for specific types of configurations that need a complex data structure (like the list of noteburst users). So I guess you'd call this pattern: foundational configuration via environment variables and defering to files for special configuration tasks.

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+    For exceptions designed to generate structured JSON bodies as part of HTTP errors, define a ``to_dict`` method that translates the exceptions into a dictionary suitable for serializing to JSON.
+
+``factory.py``
+    Contains the factory object used to construct services and their dependencies.
```
If you can, it might be nice sketch out the structure of the factory somewhere.

I guess the [RequestContext](https://github.com/lsst-sqre/times-square/blob/fff0c3311cb0efa04d846f95129f5bb517c2e091/src/timessquare/dependencies/requestcontext.py#L26) dependency I implemented in Times Square is a lightweight factory factory specific to HTTP requests. But if I implemented a true "factory" then I could have also used it in arq worker context, where instead I have a [separate factory for the services](https://github.com/lsst-sqre/times-square/blob/main/src/timessquare/worker/servicefactory.py)

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+The intended audience for the tech note is other sites or other project members trying to understand what we did, and any future maintainer of the application who needs to understand the underlying design principles and tradeoffs.
+
+Use ``DMTN`` tech notes for Science Platform components.
+Use ``SQR`` tech notes for internal applications.
```
My own take on this is that technotes on projects based out of the `lsst-sqre` GitHub org have SQR technotes while DMTN technotes are for projects hosted out of the `lsst` and `lsst-dm` organizations.

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+
+For separation of concerns, this structure allows clean separation between the code that converts an HTTP request into an internal call (the handler), the business logic that makes decisions about what should happen (the service), and the code to convert between internal data structures and external storage (the storage object).
+Each of these can change independently of the other or gain multiple implementations with minimum impact on the rest of the application.
+For example, one could add a command-line interface that takes some of the same actions as the web UI without duplicating code, since both would call into the services layer.
```
Same for introducing GraphQL interfaces, gRPC interfaces, Kafka topic handlers, and async workers.

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+
+The bulk of the handler should therefore be in the FastAPI decorator and in the parameter list.
+FastAPI generates the API documentation from that annotation, so make full use of all of the parameters that flesh out the documentation.
+Specifically, every handler should have a ``summary``, many handlers should have a ``responses`` parameter specifying their error codes and descriptions, many handlers should have a ``status_code`` parameter, and larger applications with a lot of handlers should use ``tags``.
```
Just thinking we might need to write up a standard for tags so that Squareone can merge open API specs.

rra replied:
Oh, yeah, that's a very good idea.

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+    All business logic should go into a service object instead.
+    This is a good place to put Pydantic validators that are shared by multiple models.
+
+If this application uses a SQL database for storage, the SQLAlchemy_ ORM models should go into a directory named ``schema``, and the ``__init__.py`` file for that directory should import all of the models.
```
In some cases I've renamed this to "dbschema" because "schema" can get a bit overloaded.

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+As you can see from that example, while FastAPI tries to produce good documentation from enums, it's often not clear enough and one may need to hand-craft a good description.
+
+Any field in a model that takes a limited set of values should be defined as a type inheriting from ``Enum``.
+I generally do not make the class also inherit from ``str`` and instead explicitly add ``.value`` to get the string value of an enum.
```
Can Pydantic serialize to JSON an Enum that doesn't inherit from `str`, or do you need to configure the serialization? I think this has been the main reason I've inherited from both str and Enum in my enums.

rra replied:
Yes, despite all the StackOverflow questions that imply it can't, this does work in Pydantic in most cases. There are a few cases that it can't handle (serializing dictionaries where the enum is the key, for instance), but for simple data members that are enum types, Pydantic apparently adds a custom JSON encoder by default that uses the enum value. I'm using this in Gafaelfawr for `TokenType` and never noticed that I didn't make it inherit from `str`.

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
+---------------
+
+For models that are only used internally (such as between services and storage objects), prefer dataclasses_ to Pydantic models.
+Dataclasses are much simpler and signal that none of the complex validation done by Pydantic is in play.
```
and significantly, none of the *data transformation*

jonathansick commented on index.rst:
```diff
@@ -1,29 +1,988 @@
 
-Add content here
-================
+Each release should is marked with a Git tag matching the version number (with no leading ``v``).
```
Potentially working linking to [setuptools_scm](https://github.com/pypa/setuptools_scm) because it means that the application version is the Git tag.

jonathansick reviewed (approved):
Great write up. Many sections may have been good to add to the Safir docs as an application development guide because a lot of them have become generally-adopted patterns, but we can instead link to this technote.

rra committed:
Add note about from __future__ import annotations

rra commented:
For the record, I'm very happy to move the bits of this on which we can get a general consensus into Safir. I think that would be a great evolution, and I could keep the tech note for just the bits that are too specific or personal opinion or whatever. That might be another good co-work topic.

rra committed:
Mention arq as another package used by Safir

rra committed:
Don't call missing features bugs

Co-authored-by: Jonathan Sick <jonathan@jsick.codes>

rra committed:
Mention other examples of new frontends

rra committed:
Dataclasses also signoal no transformation

rra committed:
Add a link to setuptools_scm

rra committed:
Elaborate on DMTN vs. SQR tech notes

Try to capture Frossie's recommendation.

rra committed:
Be explicit that list and sometimes dict are fine

No need to make a model or a type alias for these, but only use
dict when the types are consistent.

rra committed:
Add some method naming guidelines

rra committed:
More clarification of dictionary types

rra committed:
Note deprecated types

Also recommend the | syntax over Union in the few places where it
is appropriate.

rra committed:
Explicitly say Python 3.10 or later

Also give a general update policy for what version of Python to
use.

rra committed:
Add a section on service objects

Clarify what should be passed into methods, and document dependency
injection.

rra committed:
Document the storage layer

rra committed:
Update configuration suggestions

Use BaseSettings instead of a dataclass, and discuss some of the
issues when adding a class method such as from_file to a BaseSettings
class that also requires Pydantic configuration.  Note that the new
Union syntax requires Python 3.10.

rra committed:
General updates from a re-read

- Mention that Config may need to be Configuration
- Say explicitly that exceptions aren't models
- Improve the config_dependency example
- Clarify the purpose of the title parameter to Field
- Include another validator example for timedelta
- Suggest using safir.pydantic.normalize_datetime for datetime
- Standardize on to_ vs. as_ for conversion methods
- Document choices for Optional vs. | None
- Use start and stop for background processing, not aclose
- Use roles to link to other tech notes

rra committed:
Add a section on factories

Describe the factory pattern in more detail, along with how to
handle command-line invocations and the Gafaelfawr ProcessContext
concept.

rra committed:
Add some style guidance for modules

rra committed:
Add intersphinx for Python core modules

rra committed:
Rename Context to RequestContext

Match the naming convention of Gafaelfawr.

rra committed:
Document accepting camel-case in models

rra committed:
Recommend naming convention for Settings/Config
